### PR TITLE
Fix health probe port overwrite

### LINKED ISSUE #45482: After upgrade to 1.18.0 Request to probe app failed
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Pods do not become ready after upgrading istio to 1.18.0

2023-06-15T12:15:15.006134Z	error	Request to probe app failed: Get "http://10.255.37.140:9090/health": dial tcp 127.0.0.6:0->10.255.37.140:9090: connect: connection refused, original URL path = /app-health/marketplace/readyzapp URL path = /health

### Version

```Text
istioctl version
client version: 1.17.2
control plane version: 1.18.0
data plane version: 1.17.2 (32 proxies), 1.18.0 (5 proxies)

kubectl version --short
Client Version: v1.27.2
Kustomize Version: v5.0.1
Server Version: v1.26.3
```

### Additional Information

only the workloads running 1.18.0 proxies are failing

helm chart values for istiod:

```yaml
global:
  proxy:
    holdApplicationUntilProxyStarts: true
    resources:
      requests:
        cpu: 1m
        memory: 128Mi
      limits:
        cpu: 2000m
        memory: 1Gi
# meshConfig:
#   outboundTrafficPolicy:
#     mode: REGISTRY_ONLY
pilot:
  autoscaleEnabled: false
  replicaCount: 2
```

downgrading istio to 1.17.3 resovles the issue

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

### PR DESCRIPTION
**Pull request by jaellio**
**Please provide a description of this PR:**

Resolves #45482.

This change removes the overwrite of the host of the application request created by the istio-agent status server. Previously, the host was overwritten to the host of the original request from Kubelet to the istio-agent status server. Unless a host header was specified, the host would always be set to the status port, 15020, rather than the application port used for health checks.

Testing:
- Modified unit tests to remove explicit port override. Without the override in the unit tests, the existing workflow fails as expected. The unit test override did not reflect the actual workflow and values of a health check request.

### CODE REVIEW

File: pilot/cmd/pilot-agent/status/server.go [hunk 1]
```diff
@@ -709,10 +709,11 @@ func (s *Server) handleAppProbeHTTPGet(w http.ResponseWriter, req *http.Request,
 	}
 
 	// Forward incoming headers to the application.
-	appReq.Host = req.Host
+	appReq.Host = prober.HTTPGet.Host
```
Comment (jaellio) [inline pilot/cmd/pilot-agent/status/server.go:712]: Should this be overwritten only if `prober.HTTPGet.HOST` isn't empty?

File: pilot/cmd/pilot-agent/status/server.go [hunk 2]
```diff
@@ -709,10 +709,13 @@ func (s *Server) handleAppProbeHTTPGet(w http.ResponseWriter, req *http.Request,
 	}
 
 	// Forward incoming headers to the application.
-	appReq.Host = req.Host
+	if prober.HTTPGet.Host != "" {
+		appReq.Host = prober.HTTPGet.Host
```
Comment (hzxuzhonghu) [inline pilot/cmd/pilot-agent/status/server.go:713]: better test with the depoyment to check why connection refused

Comment (howardjohn) [inline pilot/cmd/pilot-agent/status/server.go:713]: Won't this get overriden by L719?

File: pilot/cmd/pilot-agent/status/server.go [hunk 3]
```diff
@@ -709,10 +709,13 @@ func (s *Server) handleAppProbeHTTPGet(w http.ResponseWriter, req *http.Request,
...
+	}
 	for name, values := range req.Header {
 		appReq.Header[name] = slices.Clone(values)
 		if len(values) > 0 && (strings.EqualFold(name, "Host") || name == ":authority") {
+			// Probe has specific host header override; honor it
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/status/server.go:718]: One thing I am not clear on...

isn't `req.Host == req.Header["Host"]`? So how can we have this code and its fine (I tested it -- it works great as is in the PR), but it still works.

**Review by howardjohn — APPROVED** [review]

File: pilot/cmd/pilot-agent/status/server.go (hunk 3 above)
Comment (jaellio) [reply pilot/cmd/pilot-agent/status/server.go:718]: Not sure I totally understand your question here

**Comment by jaellio** [comment]
> LGTM, I had a question though so I put a hold. Feel free to remove it

Does this fix need a release note?

File: pilot/cmd/pilot-agent/status/server.go (hunk 2 above)
Comment (jaellio) [reply pilot/cmd/pilot-agent/status/server.go:713]: I replicated the health check failing with a 404 error. I am not sure why a connection refused error was seen.

File: pilot/cmd/pilot-agent/status/server.go (hunk 3 above)
Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:718]: From Godoc:

```
	// For incoming requests, the Host header is promoted to the
	// Request.Host field and removed from the Header map.
```

So is it actually possible for this line to be hit?

Comment (jaellio) [reply pilot/cmd/pilot-agent/status/server.go:718]: Maybe not for "Host" header but for the ":authority" header?

Comment (howardjohn) [reply pilot/cmd/pilot-agent/status/server.go:718]: I would think the same applies to both of those.. maybe not though

**Pull request merged**